webhooks/virtualmachine/validation/scsi_controller.go

```diff
@@ -0,0 +1,259 @@
+//     (beyond max allowed). and there must be an entry in
+//     spec.hardware.controllers for thiscontroller.
+//   - verify that we don't have more devices than max allowed for
+//     each controller type.
+//
+// Importantly, if a volume doesn't specify a controllerBusNumber, or
+// if we can't find a controller for it, we don't reject the request.
+func (v validator) validateControllerSlots(
+	_ *pkgctx.WebhookRequestContext,
+	vm *vmopv1.VirtualMachine) field.ErrorList {
+
+	var allErrs field.ErrorList
+
+	if len(vm.Spec.Volumes) == 0 {
+		return allErrs
+	}
+
+	volumesPath := field.NewPath("spec", "volumes")
+
+	// Build controller map from spec.hardware.controllers
+	specControllers := make(map[int32]controllerInfo)
+	if vm.Spec.Hardware != nil {
+		for _, controller := range vm.Spec.Hardware.SCSIControllers {
+			specControllers[controller.BusNumber] = controllerInfo{
+				busNumber:   controller.BusNumber,
+				ctrlType:    controller.Type,
+				sharingMode: controller.SharingMode,
+				maxSlots:    getMaxSlotsForControllerType(controller.Type),
+				usedSlots:   0,
+			}
+		}
+	}
+
+	// Build device count map from status.hardware.controllers
+	statusDeviceCounts := make(map[int32]int32)
+	if vm.Status.Hardware != nil {
+		for _, controller := range vm.Status.Hardware.Controllers {
+			if controller.Type == vmopv1.VirtualControllerTypeSCSI {
+				statusDeviceCounts[controller.BusNumber] = int32(len(controller.Devices)) //nolint:gosec // disable G115
+			}
+		}
+	}
+
+	// Track volume assignments to controllers
+	volumeAssignments := make(map[int32][]string) // busNumber -> volume names
+
+	// Process each volume
+	for i, vol := range vm.Spec.Volumes {
+		if vol.PersistentVolumeClaim == nil {
+			continue
+		}
+
+		pvc := vol.PersistentVolumeClaim
+		volPath := volumesPath.Index(i)
+
+		// Determine which controller this volume should use
+		var targetBusNumber int32
+		var targetController controllerInfo
+		var hasTarget bool
+
+		// Check if a specific controller is requested
+		if pvc.ControllerBusNumber != nil {
+			targetBusNumber = *pvc.ControllerBusNumber
+
+			// Validate bus number range
+			if targetBusNumber < 0 || targetBusNumber >= maxSCSIControllers {
+				allErrs = append(allErrs, field.Invalid(
+					volPath.Child("persistentVolumeClaim", "controllerBusNumber"),
+					targetBusNumber,
+					fmt.Sprintf(invalidControllerBusNumberRangeFmt, maxSCSIControllers-1)))
+				continue
+			}
+
+			targetController, hasTarget = specControllers[targetBusNumber]
+			if !hasTarget {
+				allErrs = append(allErrs, field.Invalid(
+					volPath.Child("persistentVolumeClaim", "controllerBusNumber"),
+					targetBusNumber,
+					fmt.Sprintf(invalidControllerBusNumberDoesNotExist, targetBusNumber)))
+				continue
+			}
+		} else {
+			// Find appropriate controller based on application type
+			switch pvc.ApplicationType {
+			case vmopv1.VolumeApplicationTypeOracleRAC:
+				targetController, hasTarget = findControllerWithSharing(
```

Comment:
When choosing the controller, we didn't consider the maximum slots a bus can have? Wonder would it cause issue in VMOP controller.